File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1 for _ in range(n)]
        ans = 0
        def recurse(index):
            if index==n-2 or index==n-1:
                dp[index] = nums[index]
                return dp[index]
            if dp[index]!=-1:
                return dp[index]
            for i in range(index+2,n):
                res = recurse(i)
                dp[index] = max(nums[index]+res,dp[index])
            return dp[index]
        #print(dp)
        for j in range(n):
            if dp[j]==-1:
                tmp = recurse(j)
                ans = max(tmp,ans)
        return ans

File: 198-house-robber/198-house-robber.java
class Solution {
    int dp[];
    public int helper(int[] nums, int index){
        if(index==nums.length-2 || index==nums.length-1){
            dp[index] = nums[index];
            return nums[index];
        }
        if(dp[index]!=-1){
            return dp[index];
        }
        for(int i=index+2;i<nums.length;i++){       
            dp[index] = Math.max(nums[index]+helper(nums,i),dp[index]); //check if the current steal+steal from other houses is greater than the steal specified in the dp array
        }
        return dp[index];
    }
    public int rob(int[] nums) {
        if(nums.length==1){
            return nums[0];
        }
        dp = new int[nums.length];
        int ans = 0;
        for(int i=0;i<nums.length;i++){ //initialize the dp array to '-1' each
            dp[i] = -1;
        }
        for(int j=0;j<nums.length;j++){ //with different starting points
            if(dp[j]==-1){
            ans = Math.max(helper(nums,j),ans);
            }
        }
        return ans;
    }
}